src/pudl/output/ferc1.py

```diff
@@ -2774,3 +2779,239 @@ def nodes_to_df(calc_forest: nx.DiGraph, nodes: list[NodeId]) -> pd.DataFrame:
+            for desc in descendants
+        }
+        nx.set_node_attributes(annotated_forest, descendant_tags)
+    return annotated_forest
+
+
+def _propagate_tag_rootward(
+    annotated_forest: nx.DiGraph, node, tag_name: Literal["in_rate_base"]
+) -> str:
+    """Set the tag for nodes when all of its children have same tag.
+
+    This function returns the value of a tag, but also sets node attributes
+    down the tree when all children of a node share the same tag.
+    """
+
+    def _get_tag(annotated_forest, node, tag_name):
+        return annotated_forest.nodes.get(node, {}).get("tags", {}).get(tag_name, pd.NA)
+
+    logger.info(f"propagaging tags leafward from {node}")
+    tag = pd.NA
+    # i'm a leaf so i stop looking or
+    # if i have a value you don't need to keep looking at this node's childern
+    if not list(annotated_forest.successors(node)) or not pd.isna(
+        _get_tag(annotated_forest, node, tag_name)
+    ):
+        tag = _get_tag(annotated_forest, node, tag_name)
+
+    else:
+        child_tags = set()
+        for child_node in annotated_forest.successors(node):
+            if not child_node.xbrl_factoid.endswith("_correction"):
+                child_tags.add(
+                    _propagate_tag_rootward(annotated_forest, child_node, tag_name)
+                )
+        # if all the children tags are the same and non-null
+        if (len(child_tags) == 1) and {t for t in child_tags if not pd.isna(t)}:
+            new_node_tag = child_tags.pop()
+            # actually assign the tag here but don't wipe out any other tags
+            existing_tags = nx.get_node_attributes(annotated_forest, "tags")
+            node_tags = {
+                node: {"tags": {tag_name: new_node_tag} | existing_tags.get(node, {})}
+            }
+            nx.set_node_attributes(annotated_forest, node_tags)
+            tag = new_node_tag
+        # elif the children disagree then the node's tag shouldn't be set and
+        # the og null tag should be returned
+    return tag
+
+
+def _propagate_tags_to_corrections(annotated_forest: nx.DiGraph) -> nx.DiGraph:
+    existing_tags = nx.get_node_attributes(annotated_forest, "tags")
+    correction_nodes = [
+        node for node in annotated_forest if node.xbrl_factoid.endswith("_correction")
+    ]
+    correction_tags = {}
+    for correction_node in correction_nodes:
+        # for every correction node, we assume that that nodes parent tags can apply
+        parents = list(annotated_forest.predecessors(correction_node))
+        # all correction records shoul have a parent and only one
+        assert len(parents) == 1
+        parent = parents[0]
+        correction_tags[correction_node] = {
+            "tags": existing_tags.get(parent, {})
+            | existing_tags.get(correction_node, {})
+        }
+    nx.set_node_attributes(annotated_forest, correction_tags)
+    return annotated_forest
+
+
+def check_tag_propagation_compared_to_compiled_tags(
```

Comment:
i added two `check_*` functions here. they are both being called rn in `out_ferc1__yearly_rate_base`. really these should be validation tests i believe for both the rate base table and the two exploded tables detailed in the args. i could put it into the `exploded_table_asset_factory` and just skip it for the income table. thoughts?

Reply:
Calling them in out_ferc1__yearly_rate_base seems fine, though you could try stuffing them in an [asset check](https://docs.dagster.io/concepts/assets/asset-checks) in the factory. I think either is fine.